julia_src/reopt_model.jl

```diff
@@ -1145,8 +1145,202 @@ function add_cost_function(m, p)
     #= Note: 0.9999*m[:MinChargeAdder] in Obj b/c when m[:TotalMinCharge] > (TotalEnergyCharges + m[:TotalDemandCharges] + TotalExportBenefit + m[:TotalFixedCharges])
 		it is arbitrary where the min charge ends up (eg. could be in m[:TotalDemandCharges] or m[:MinChargeAdder]).
 		0.0001*m[:MinChargeAdder] is added back into LCC when writing to results.  =#
+
+	# Add climate costs
+	if p.Include_climate_in_objective # if user selects to include climate in objective
+		add_to_expression!(Costs, m[:Lifecycle_Emissions_Cost_CO2])
+	end
+
+	# Add Health costs (NOx, SO2, PM2.5)
+	if p.Include_health_in_objective
+		add_to_expression!(Costs, m[:Lifecycle_Emissions_Cost_Health])
+	end
+
+end
+
+# Renewable electricity calculation
+function add_re_elec_calcs(m,p)
+	if p.IncludeExportedREElecinTotal
+		include_exported_re_elec_in_total = 1
+	else
+		include_exported_re_elec_in_total = 0
+	end
+	m[:AnnualREEleckWh] = @expression(m,p.TimeStepScaling*
+		(sum(p.ProductionFactor[t,ts] * p.LevelizationFactor[t] * m[:dvRatedProduction][t,ts] * p.TechPercentRE[t] for t in p.ElectricTechs, ts in p.TimeStep) #total RE elec generation, excl steam turbine
+		- sum(m[:dvProductionToStorage][b,t,ts]*p.TechPercentRE[t]*(1-p.ChargeEfficiency[t,b]*p.DischargeEfficiency[b]) for t in p.ElectricTechs, b in p.ElecStorage, ts in p.TimeStep) #minus battery efficiency losses
+		- sum(m[:dvProductionToCurtail][t,ts]*p.TechPercentRE[t] for t in p.ElectricTechs, ts in p.TimeStep) # minus curtailment.
+		- (1-include_exported_re_elec_in_total)*sum(m[:dvProductionToGrid][t,u,ts]*p.TechPercentRE[t] for t in p.ElectricTechs,  u in p.ExportTiersByTech[t], ts in p.TimeStep) # minus exported RE, if RE accounting method = 0.
+		+ p.STElecOutToThermInRatio * sum(m[:dvThermalToSteamTurbine][tst,ts]*p.TechPercentRE[tst] for ts in p.TimeStep, tst in p.TechCanSupplySteamTurbine) )) # plus steam turbine generation (steam turbine p.TechPercentRE input = 0 because it is a decision variable dependent on fraction of steam generated by RE fuel
+		# TODO: Steam Turbine battery losses and curtailment (challenge is that equations seem non-linear)
```

Comment:
I agree if ST can curtail, this gets tricky. However, for the immediate need of emissions in the webtool, ST defaults to **cannot** curtail, and we don't expose that input in the webtool. So ST will never curtail in a webtool analysis (same with CHP). Also, only **thermal**-producing techs (and only Boiler in the webtool) `can_supply_steam_turbine` - and curtailment only applies to the electricity. I would also generally advise against ever letting ST or CHP curtail, even for API usage (and maybe that points to removing `can_curtail` from the inputs of those Techs, but there may be cases to model them). Does this resolve this issue? Or maybe there's still the issue that in the API, Boiler, NewBoiler and/or CHP `can_supply_steam_turbine`, and each of them may have different %RE of their fuel. In that case, the %RE for ST should be the weighted average of `dvThermalToSteamTurbine[tst] * p.TechPercentRE[tst] for tst in p.TechCanSupplySteamTurbine`.

Reply:
Even if ST can't curtail, the problem still remains for energy going from ST to storage. If you calculate the ST %RE as that weighted average and then multiply it by the DV for ST to storage, that not linear.

Reply:
I see that now, thanks. I agree we should document somewhere that we currently cannot accurately model emissions for scenarios where multiple `HeatingTechs` have `can_supply_steam_turbine` = True **and** the `HeatingTechs` have different %-RE fuels. We don't need to spend time figuring a solution to that now. The Webtool will not be able to model that scenario, and it won't be such a common use case in general, for the API users.

Reply:
General question: are we not counting energy that goes through `Storage` as renewable, not matter the source of it?

Reply:
The documentation should say that we do not accurately model emissions and RE% for scenarios where _any_ techs can supply steam turbine. Yes, if there was only one then we could model it, but to do that we would need to put in some "if only one tech supplies ST, do this" code. 
I think we are counting renewable energy that goes through storage because storage can't export to grid so that's not an issue.

Reply:
@hdunham I see now that we are just subtracting the battery losses, not the whole `dvProductionToStorage`. I'm still a little confused about why SteamTurbine accounting can't be done. Please ping me when you get a chance to chat.

Reply:
Hi @Bill-Becker and @hdunham , to confirm, both of the following calculations would be non-linear and thus cause problems, right?

(Option 1) calculate the overall weighted average percentage of input steam to SteamTurbine that is RE (using equation below, which I think is nonlinear, right?), and apply that to dvProductionToStorage for SteamTurbineTechs, in replacement of the non-weighted average approximation:

{weighted average percent RE steam feeding SteamTurbine} =
sum(m[:dvThermalToSteamTurbine][tst,ts]*p.TechPercentRE[tst] for ts in p.TimeStep, tst in p.TechCanSupplySteamTurbine) / 
sum(m[:dvThermalToSteamTurbine][tst,ts] for ts in p.TimeStep, tst in p.TechCanSupplySteamTurbine) 

(Option 2) Calculate the fraction of steam turbine elec generation that goes to storage (see below) and subtract that portion out from the line of code above?:

{fraction of steam turbine elec gen going to storage} = 
sum(m[:dvProductionToStorage][b,t,ts] for b in p.ElecStorage, t in p.SteamTurbineTechs, ts in p.TimeStep) / sum(m[:dvRatedProduction][t,ts]  for t in p.SteamTurbineTechs, ts in p.TimeStep)

then could do something like 
 p.STElecOutToThermInRatio * (1-{fraction of steam turbine elec gen going to storage}*{storage losses})
 
If these are both non-linear, which I think they are, one other idea that comes to mind is that non-linear calculations should be okay for post-processing, just not the optimization, so perhaps we could provide accurate post-processed calculations and either prohibit users from including the SteamTurbine in any scenarios with RE elec targets or just use the approximation approach, with a caveat, if RE targets are set with a SteamTurbine tech? Though I suppose that could cause more confusion than it alleviates, especially since this is such a specific and probably unlikely use case...

Reply:
Also @hdunham - just want to confirm, if `if isempty(p.SteamTurbineTechs)` is true, we can delete the line 

"        + p.STElecOutToThermInRatio * sum(m[:dvThermalToSteamTurbine][tst,ts]*p.TechPercentRE[tst] for ts in p.TimeStep, tst in p.TechCanSupplySteamTurbine) )) # plus steam turbine generation (steam turbine p.TechPercentRE input = 0 because it is a decision variable dependent on fraction of steam generated by RE fuel "

correct?

![image](https://user-images.githubusercontent.com/35976768/144178483-920fcf31-f3b2-41b7-bade-91809264ef7b.png)

Reply:
Okay and one more comment related to this topic... another, possibly simpler (but more conservative) alternative to the non-weighted averaging approach could be to conservatively set p.TechPercentRE for SteamTurbine techs to be the min of all p.TechPercentREs for all techs in p.TechCanSupplySteamTurbine- this could be done in data_manager.py instead of complicating the julia code. But as mentioned would then underestimate the RE elec generation

Reply:
@kkrah, @hdunham and I discussed this yesterday and thought to just use the **average** p.TechPercentRE of all p.TechCanSupplySteamTurbine for the battery storage losses. This will be perfectly accurate if all p.TechCanSupplySteamTurbine have the same p.TechPercentRE (including if there is only **one** tech in p.TechCanSupplySteamTurbine which will be the case for the Webtool currently), and it will be inaccurate if they don't have the same p.TechPercentRE **and** more than one techs are supplying SteamTurbine **and** SteamTurbine is charging the battery. But, since this is just for the battery **losses** for SteamTurbine, and we don't think SteamTurbine will charge the battery much, we think this is fine. We could also do your suggestion to accurately calculate it in the post-processing, even if the optimization uses the average (possibly inaccurate) for ST battery losses. We should add to the documentation that "there may be small inaccuracy in RE-% targets for scenarios where multiple [HeatingTechs] have [can_supply_steam_turbine = True] and the [HeatingTechs] have different %-RE fuels.

Reply:
@kkrah Also, I think @hdunham plans to push this implementation soon.

Reply:
@Bill-Becker awesome, sounds great to me!! thanks!

Reply:
Bill, I pushed it the day we discussed. Kathleen, yes that line is not necessary. And I see you also subtracted the ST export term, thanks for catching that.